### PR TITLE
Atmosphere Emoji

### PR DESCRIPTION
**Pull request by dmitry-yudakov**
We're loading emojis config for the world and showing the Emoji bar on down-left position.

When clicking on an emoji icon a relay message is sent to Unity - if we're flying around the world, it will show the animation.

If we're in some Space (collaboration module), we need to show the animation ourselves. For that a special EmojiAnimationDock component is created, allowing to handle events from unity and show Emoji or Megamoji animation.

### CODE REVIEW

File: src/scenes/collaboration/Collaboration.styled.ts [hunk 1]
```diff
@@ -4,3 +4,15 @@ export const Container = styled.ul`
...
+export const BottomCenteredDock = styled.div`
+  position: fixed;
+  bottom: 50px;
+  top: 100px;
+  left: 50%;
+  transform: translateX(-50%);
+  width: 20%;
+  z-index: calc(var(--overlay-z-index) + 1);
+  pointer-events: none;
+  // background: #ff000070;
```
Comment (PAKovalev) [inline src/scenes/collaboration/Collaboration.styled.ts:17]: remove

File: src/scenes/collaboration/components/EmojiAnimationDock/EmojiAnimationDock.tsx [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+import {usePosBusEvent} from 'shared/hooks';
+import {PosBusEmojiMessageType} from 'core/types';
+
+import {EmojiWithAvatarAnimation, MegamojiAnimation} from './components';
+
+const ANIMATION_DURATION_SEC = 5;
+
+interface EmojiItemInterface {
+  type: 'emoji' | 'megamoji';
```
Comment (PAKovalev) [inline src/scenes/collaboration/components/EmojiAnimationDock/EmojiAnimationDock.tsx:12]: Make a type under `core/types`

File: src/scenes/collaboration/components/EmojiAnimationDock/EmojiAnimationDock.tsx [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        type,
+        emojiUrl: typeof emojiUrlOrMsg === 'string' ? (emojiUrlOrMsg as string) : null,
+        emojiMsg:
+          typeof emojiUrlOrMsg === 'object' ? (emojiUrlOrMsg as PosBusEmojiMessageType) : null,
+        idx: refIndex.current++,
+        offset: Math.ceil(Math.random() * 100)
+      };
+
+      setItems((items) => new Set(items).add(item));
+      // console.log('MOUNT', item);
```
Comment (PAKovalev) [inline src/scenes/collaboration/components/EmojiAnimationDock/EmojiAnimationDock.tsx:36]: remove

File: src/scenes/collaboration/components/EmojiAnimationDock/components/EmojiWithAvatarAnimation/EmojiWithAvatarAnimation.styled.ts [hunk 4]
```diff
@@ -0,0 +1,135 @@
+import styled from 'styled-components';
+
+export const EmojiContainer = styled.div`
+  position: absolute;
+  // position: fixed;
```
Comment (PAKovalev) [inline src/scenes/collaboration/components/EmojiAnimationDock/components/EmojiWithAvatarAnimation/EmojiWithAvatarAnimation.styled.ts:5]: remove

File: src/scenes/collaboration/components/EmojiAnimationDock/components/EmojiWithAvatarAnimation/EmojiWithAvatarAnimation.styled.ts [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+  }
+`;
+
+export const EmojiImg = styled.img`
+  position: absolute;
+  animation-name: emoji;
+  animation-duration: 3s;
+  opacity: 0;
+  animation-timing-function: linear;
+  // animation-iteration-count: infinite;
```
Comment (PAKovalev) [inline src/scenes/collaboration/components/EmojiAnimationDock/components/EmojiWithAvatarAnimation/EmojiWithAvatarAnimation.styled.ts:45]: remove

File: src/scenes/collaboration/components/EmojiAnimationDock/components/EmojiWithAvatarAnimation/EmojiWithAvatarAnimation.styled.ts [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+  }
+`;
+
+export const Avatar = styled.div`
+  position: absolute;
+  animation-name: avatar;
+  animation-duration: 3s;
+  animation-timing-function: linear;
+  opacity: 0;
+  // animation-iteration-count: infinite;
```
Comment (PAKovalev) [inline src/scenes/collaboration/components/EmojiAnimationDock/components/EmojiWithAvatarAnimation/EmojiWithAvatarAnimation.styled.ts:89]: remove

File: src/scenes/widgets/pages/EmojiWidget/EmojiWidget.tsx [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+import {EmojiItem} from './components/';
+
+interface PropsInterface {
+  onClose: () => void;
+}
+
+const EmojiWidget: FC<PropsInterface> = ({onClose}) => {
+  const {
+    widgetStore,
+    // mainStore,
```
Comment (PAKovalev) [inline src/scenes/widgets/pages/EmojiWidget/EmojiWidget.tsx:19]: remove

**Review by PAKovalev — APPROVED** [review]

File: src/scenes/collaboration/components/EmojiAnimationDock/EmojiAnimationDock.tsx (hunk 2 above)
Comment (dmitry-yudakov) [reply src/scenes/collaboration/components/EmojiAnimationDock/EmojiAnimationDock.tsx:12]: It's pretty internal thing to this component, I doubt we should pollute global space with them?

File: src/scenes/collaboration/components/EmojiAnimationDock/EmojiAnimationDock.tsx [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+import {observer} from 'mobx-react-lite';
+
+import {usePosBusEvent} from 'shared/hooks';
+import {PosBusEmojiMessageType} from 'core/types';
+
+import {EmojiWithAvatarAnimation, MegamojiAnimation} from './components';
+
+const ANIMATION_DURATION_SEC = 5;
+
+interface EmojiItemInterface {
```
Comment (aminbksi) [inline src/scenes/collaboration/components/EmojiAnimationDock/EmojiAnimationDock.tsx:11]: move to interfaces

File: src/scenes/collaboration/components/EmojiAnimationDock/EmojiAnimationDock.tsx [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  type: 'emoji' | 'megamoji';
+  emojiUrl: string | null;
+  emojiMsg: PosBusEmojiMessageType | null;
+  idx: number;
+  offset: number;
+}
+type EmojiItemInterfaceStoreType = Set<EmojiItemInterface>;
```
Comment (aminbksi) [inline src/scenes/collaboration/components/EmojiAnimationDock/EmojiAnimationDock.tsx:18]: move to types

File: src/scenes/collaboration/components/EmojiAnimationDock/EmojiAnimationDock.tsx [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+const EmojiAnimationDock: FC = () => {
+  const [items, setItems] = useState<EmojiItemInterfaceStoreType>(() => new Set());
+  const refIndex = useRef(1);
+
+  const newEmojiHandler =
```
Comment (aminbksi) [inline src/scenes/collaboration/components/EmojiAnimationDock/EmojiAnimationDock.tsx:24]: move to store

File: src/scenes/collaboration/components/EmojiAnimationDock/components/EmojiWithAvatarAnimation/EmojiWithAvatarAnimation.tsx [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+import {PosBusEmojiMessageType} from 'core/types';
+
+import * as styled from './EmojiWithAvatarAnimation.styled';
+
+interface PropsInterface {
+  emojiMsg: PosBusEmojiMessageType;
+  offset: number;
+}
+
+const createLettersAvatar = (name: string): string => {
```
Comment (aminbksi) [inline src/scenes/collaboration/components/EmojiAnimationDock/components/EmojiWithAvatarAnimation/EmojiWithAvatarAnimation.tsx:12]: move to utils

File: src/scenes/widgets/Widgets.tsx [hunk 12]
```diff
@@ -51,7 +53,8 @@ const Widgets: FC = () => {
 
   useEffect(() => {
     musicPlayerStore.init(worldStore.worldId);
-  }, [musicPlayerStore, worldStore.worldId]);
+    emojiStore.init(worldStore.worldId);
```
Comment (aminbksi) [inline src/scenes/widgets/Widgets.tsx:56]: why you put in the same hook as musicPlayerStore?

File: src/scenes/widgets/pages/EmojiWidget/EmojiWidget.tsx [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+const EmojiWidget: FC<PropsInterface> = ({onClose}) => {
+  const {widgetStore, sessionStore} = useStore();
+  const {emojiDetailsList} = widgetStore.emojiStore;
+  const {profile} = sessionStore;
+
+  const {t} = useTranslation();
+
+  const userUUID = profile?.uuid;
```
Comment (aminbksi) [inline src/scenes/widgets/pages/EmojiWidget/EmojiWidget.tsx:23]: we usually do not make constant like this, maybe Daniel and Andrey could also comment on this

File: src/scenes/widgets/stores/EmojiStore/EmojiStore.tsx [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+      fetchSpaceEmojisRequest: types.optional(RequestModel, {})
+    })
+  )
+  .actions((self) => ({
+    init: flow(function* (worldId: string) {
+      const response: EmojiConfigResponse = yield self.fetchSpaceEmojisRequest.send(
+        api.emojiRepository.fetchSpaceEmojiConfig,
+        {worldId}
+      );
+      console.log('EMOJIS response:', response);
```
Comment (aminbksi) [inline src/scenes/widgets/stores/EmojiStore/EmojiStore.tsx:22]: remove console logs

File: src/shared/services/i18n/locales/en-gb.ts [hunk 15]
```diff
@@ -179,6 +179,7 @@ export const enGb = {
       screenShare: 'Screenshare',
       stageMode: 'Stage Mode',
       calendar: 'World Calendar',
+      emoji: 'EMOJI',
```
Comment (aminbksi) [inline src/shared/services/i18n/locales/en-gb.ts:182]: use css for text transformation

File: src/scenes/widgets/pages/EmojiWidget/EmojiWidget.tsx (hunk 13 above)
Comment (Daboo1998) [reply src/scenes/widgets/pages/EmojiWidget/EmojiWidget.tsx:23]: It's not wrong, but I would say unnecessary, cause, its not much shorter typing `userUUID`, than `profile.uuid`.

File: src/scenes/widgets/Widgets.tsx (hunk 12 above)
Comment (Daboo1998) [reply src/scenes/widgets/Widgets.tsx:56]: Its not wrong, as both are dependent on `worldStore.worldId`, so putting those together is alright @aminbksi

File: src/scenes/widgets/pages/EmojiWidget/EmojiWidget.tsx [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+      }
+    },
+    [userUUID, userAvatarSrc, userName]
+  );
+
+  return (
+    <PanelLayout title={t('labels.emoji')} onClose={onClose}>
+      <styled.Container>
+        <styled.EmojiList>
+          {emojiDetailsList.map((em) => (
```
Comment (Daboo1998) [inline src/scenes/widgets/pages/EmojiWidget/EmojiWidget.tsx:41]: `em` name is not descriptive at all, I would name it as what it is, an `emoji`

File: src/shared/services/i18n/locales/en-gb.ts (hunk 15 above)
Comment (Daboo1998) [reply src/shared/services/i18n/locales/en-gb.ts:182]: +1

File: src/shared/services/unity/UnityService.tsx [hunk 17]
```diff
@@ -145,6 +146,47 @@ export class UnityService {
...
+    userName,
+    targetType = UNITY_TARGET_TYPE.USER
+  }: {
+    emojiId: string;
+    emojiUrl: string;
+    userUUID: string;
+    userAvatarSrc: string;
+    userName: string;
+    targetType?: UNITY_TARGET_TYPE;
+  }) {
```
Comment (Daboo1998) [inline src/shared/services/unity/UnityService.tsx:163]: I would make an interface for it

File: src/shared/services/unity/UnityService.tsx [hunk 18]
```diff
@@ -145,6 +146,47 @@ export class UnityService {
...
+    try {
+      console.log('SEND EMOJI:', {
+        targetType,
+        userUUID,
+        userAvatarSrc,
+        userName,
+        emojiUrl,
+        emojiId
+      });
+      console.log('unityAPI', this.unityApi);
```
Comment (Daboo1998) [inline src/shared/services/unity/UnityService.tsx:173]: Remove logs

File: src/shared/services/unity/UnityService.tsx [hunk 19]
```diff
@@ -145,6 +146,47 @@ export class UnityService {
...
+        targetType,
+        targetID: userUUID,
+        urlAvatar: userAvatarSrc,
+        nickname: userName,
+        url: emojiUrl,
+        emojiID: emojiId
+      };
+      this.unityApi?.relayMessage(topic, JSON.stringify(data));
+    } catch (error) {
+      console.error(error);
```
Comment (Daboo1998) [inline src/shared/services/unity/UnityService.tsx:186]: maybe throw it and/or handle it with a toast?

File: src/scenes/collaboration/components/EmojiAnimationDock/EmojiAnimationDock.tsx (hunk 8 above)
Comment (dmitry-yudakov) [reply src/scenes/collaboration/components/EmojiAnimationDock/EmojiAnimationDock.tsx:11]: It's an interface that is internal to this component, no point of polluting the global space with it.

File: src/shared/services/unity/UnityService.tsx (hunk 19 above)
Comment (dmitry-yudakov) [reply src/shared/services/unity/UnityService.tsx:186]: maybe, but it's a common problem for this whole UnityService file - we should perhaps review and refactor it regarding the error handling. In this case specifically failed emoji sending seems not really worth a toast...

**Pull request merged**